metrics/metrics.go

```diff
@@ -1,92 +1,183 @@
+
+	"github.com/sirupsen/logrus"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter/noop"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter/prometheus"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter/statsd"
 )
 
-// MetricCollector interface to handle metric counters
+// MonitoringConfig for profiler and prometheus
+type MonitoringConfig struct {
+	// PrometheusMetricsPort port to run prometheus on
+	PrometheusMetricsPort int `json:"prometheus_metrics_port,omitempty"`
+
+	// Statsd metrics if you are not using prometheus
+	Statsd *StatsdConfig `json:"statsd,omitempty"`
+
+	// ProfilerPort if non zero enable http profiler om this port
+	ProfilerPort int `json:"profiler_port,omitempty"`
+
+	// ProfilingPath is the variable that enable deep profiling is set
+	ProfilingPath string `json:"profiling_path,omitempty"`
+
+	ProfilerFile *os.File
+}
+
+// StatsdConfig for metrics
+type StatsdConfig struct {
+	// HostPort host:port of the statsd server
+	HostPort string `json:"host,omitempty"`
+
+	// StatsdPrefix prefix for statsd metrics
+	Prefix string `json:"prefix,omitempty"`
+
+	// StatsFlushPeriod in ms
+	FlushPeriod int `json:"flush_period,omitempty"`
+}
+
+// MetricCollector provides means to create new collectors
 type MetricCollector interface {
-	Increment(bucket string)                 // increment a bucket
-	Count(bucket string, n interface{})      // add a number to a bucket
-	Gauge(bucket string, n interface{})      // gauge a bucket
-	Timing(bucket string, value interface{}) // add a timing value
+	RegisterCounter(adapter.CollectorOptions) adapter.Counter
+	RegisterGauge(adapter.CollectorOptions) adapter.Gauge
+	RegisterTimer(adapter.CollectorOptions) adapter.Timer
+	Shutdown()
 }
 
-// StatsIncrement Deals with the statsd client to increment metric counters
-func StatsIncrement(statsCollector MetricCollector, statKey string, value int64, labels map[string]string) {
-	sendMetrics(statsCollector, statKey, value, labels, statsCollector.Count)
+// NewCollector creates a collector based on monitoring configuration
+func NewCollector(monitoringConfig *MonitoringConfig, logger *logrus.Logger) MetricCollector {
+	if monitoringConfig == nil || (monitoringConfig.PrometheusMetricsPort == 0 && monitoringConfig.Statsd == nil) {
+		logger.Infoln("config_skipping_empty_metrics_provider")
+		return noop.NewCollector()
+	}
+
+	if monitoringConfig.PrometheusMetricsPort > 0 {
+		return prometheus.NewCollector()
+	} else if monitoringConfig.Statsd != nil {
+		flushDuration := 100 * time.Millisecond
+		if monitoringConfig.Statsd.FlushPeriod > 0 {
+			flushDuration = time.Duration(monitoringConfig.Statsd.FlushPeriod) * time.Millisecond
+		}
+		collector := statsd.NewCollector(monitoringConfig.Statsd.HostPort, monitoringConfig.Statsd.Prefix, logger, flushDuration)
+		return collector
+	}
+
+	logger.Infoln("config_skipping_empty_metrics_provider")
+	return noop.NewCollector()
```

Comment:
Seems unnecessary to use the guard version and the fallback version.  Can we delete one of the two?

Reply:
Our friendly Go compiler doesn't pickup on the fact that all possible paths are taken if we remove the fallback case.

What do you think of this for maximizing readability?

```go
func NewCollector(monitoringConfig *MonitoringConfig, logger *logrus.Logger) MetricCollector {
	isPrometheus := monitoringConfig != nil && monitoringConfig.PrometheusMetricsPort > 0
	isStatsd := monitoringConfig != nil && monitoringConfig.Statsd != nil

	if isPrometheus {
		return prometheus.NewCollector()
	} 

	if isStatsd {
		flushDuration := 100 * time.Millisecond
		if monitoringConfig.Statsd.FlushPeriod > 0 {
			flushDuration = time.Duration(monitoringConfig.Statsd.FlushPeriod) * time.Millisecond
		}
		return statsd.NewCollector(monitoringConfig.Statsd.HostPort, monitoringConfig.Statsd.Prefix, logger, flushDuration)
	}

	logger.Infoln("config_skipping_empty_metrics_provider")
	return noop.NewCollector()
}
```

Reply:
looks good